lib/Modules/HyperVCIHooks/HyperVCIHooks.psm1

```diff
@@ -0,0 +1,1482 @@
+    
+    Start-ExternalCommand { cmd.exe /c "sc triggerinfo ovs-vswitchd start/strcustom/6066F867-7CA1-4418-85FD-36E3F9C0600C/VmmsWmiEventProvider" } -ErrorMessage "Failed to modify ovs-vswitchd service."
+    Start-ExternalCommand { cmd.exe /c "sc config ovs-vswitchd start=demand" } -ErrorMessage "Failed to modify ovs-vswitchd service."
+
+    Enable-OVSExtension
+
+    Enable-Service "ovsdb-server"
+    #Enable-Service "ovs-vswitchd"
+
+    Start-Service "ovsdb-server"
+    #Start-Service "ovs-vswitchd"
+}
+
+
+function Ensure-InternalOVSInterfaces {
+    $ifIndex = Get-CharmState -Namespace "novahyperv" -Key "dataNetworkIfindex"
+    $lip = Get-CharmState -Namespace "novahyperv" -Key "local_ip"
+    $ifName = (Get-NetAdapter -ifindex $ifIndex).Name
+    $lip_mask = Get-CharmState -Namespace "novahyperv" -Key "local_ip_mask"
+
+    Invoke-JujuCommand -Command @($ovs_vsctl, "--may-exist", "add-br", "juju-br")
+    Invoke-JujuCommand -Command @($ovs_vsctl, "--may-exist", "add-port", "juju-br", $ifName)
+
+    Restart-Service "ovs-vswitchd"
+    Enable-NetAdapter -Name "juju-br" -Confirm:$false
+
+    New-NetIPAddress -interfacealias "juju-br" -AddressFamily "ipv4" -IPAddress $lip -PrefixLength $lip_mask
+    Set-CharmState -Namespace "novahyperv" -Key "local_ip" -Value $lip
+    return
+}
+ 
+
+function Get-CherryPicksObject {
+    $cfgOption = Get-JujuCharmConfig -Scope 'cherry-picks'
+    if (!$cfgOption) {
+        return @{}
+    }
+    $ret = @{
+        'nova' = @();
+        'networking-hyperv' = @();
+        'compute-hyperv' = @();
+        'neutron' = @();
+        'os-win' = @()
+    }
+    $splitCfgOption = $cfgOption.Split(',')
+    $validProjects = @('nova', 'networking-hyperv', 'neutron', 'compute-hyperv', 'os-win')
+    foreach ($item in $splitCfgOption) {
+        $splitItem = $item.Split('|')
+        if ($splitItem.Count -ne 4) {
+            Throw "ERROR: Wrong 'cherry-picks' config option format"
+        }
+        $projectName = $splitItem[0]
+        if ($projectName -notin $validProjects) {
+            Throw "ERROR: Invalid git project name '$projectName'"
+        }
+        $ret[$projectName] += @{
+            'git_url' = $splitItem[1];
+            'branch_name' = $splitItem[2];
+            'commit_id' = $splitItem[3]
+        }
+    }
+    return $ret
+}
+
+
+function Initialize-GitRepository {
+    Param(
+        [string]$BuildFolder,
+        [string]$GitURL,
+        [string]$BranchName,
+        [array]$CherryPicks=@()
+    )
+
+    Write-JujuLog "Cloning $GitURL from $BranchName"
+    Start-ExecuteWithRetry { Start-GitClonePull $BuildFolder $GitURL $BranchName }
+    foreach ($commit in $CherryPicks) {
+        Write-JujuLog ("Cherry-picking commit {0} from {1}, branch {2}" -f
+                       @($commit['commit_id'], $commit['git_url'], $commit['branch_name']))
+        try {
+            Start-ExternalCommand {
+                git -C $BuildFolder fetch $commit['git_url'] $commit['branch_name']
```

Comment:
I see no test on branch here. How is it supposed to cherry-pick a commit on "master" branch if the patch under test is from "stable/newton", for instance?